# Understanding the data

First pass over the Telco churn file: **shape**, **numeric summaries**, **data quality** (missing and duplicate rows), **column names** and **target class mix**.

## Setup

Import **pandas**, **_setup** adds the project root to `sys.path`; **DATA_PATH** (from `src.config`) is the path to the CSV.

In [25]:
import pandas as pd

import _setup
from src.config import DATA_PATH

print('loaded all dependencies sucessfully...')

loaded all dependencies sucessfully...


## Load and coerce `TotalCharges`

Read the dataset with **pd.read_csv(DATA_PATH)**. Convert **TotalCharges** to numeric for numeric summary ahead

In [26]:
df = pd.read_csv(DATA_PATH)
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')
print('loaded dataset sucessfully...')

loaded dataset sucessfully...


## Shape

**df.shape** — number of rows and columns.

In [27]:
df.shape

(7043, 21)

## Numeric summary

**df.describe()** — count, mean, spread, and quantiles for numeric columns.

In [28]:
df.describe()

,SeniorCitizen,tenure,MonthlyCharges,TotalCharges
count,7043.000000,7043.000000,7043.000000,7032.000000
mean,0.162147,32.371149,64.761692,2283.300441
std,0.368612,24.559481,30.090047,2266.771362
min,0.000000,0.000000,18.250000,18.800000
25%,0.000000,9.000000,35.500000,401.450000
50%,0.000000,29.000000,70.350000,1397.475000
75%,0.000000,55.000000,89.850000,3794.737500
max,1.000000,72.000000,118.750000,8684.800000


## Missing values

Show only columns that have **at least one** missing value and **how many** 

In [29]:
df.isnull().sum()[df.isnull().sum() > 0]

TotalCharges    11
dtype: int64

## Duplicate rows

**df.duplicated().sum()** — count of fully duplicated rows.

In [30]:
df.duplicated().sum()

np.int64(0)

## Column names

**df.columns** — all variable names in the table.

In [31]:
df.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='str')

## Low-cardinality columns

For each column with **at most 10** distinct values, print **unique categories**

In [32]:
cols = [col for col in df.columns.to_list() if df[col].nunique() <= 10]
for col in cols:
    print(f'{col} : {list(df[col].unique())}')

gender : ['Female', 'Male']
SeniorCitizen : [np.int64(0), np.int64(1)]
Partner : ['Yes', 'No']
Dependents : ['No', 'Yes']
PhoneService : ['No', 'Yes']
MultipleLines : ['No phone service', 'No', 'Yes']
InternetService : ['DSL', 'Fiber optic', 'No']
OnlineSecurity : ['No', 'Yes', 'No internet service']
OnlineBackup : ['Yes', 'No', 'No internet service']
DeviceProtection : ['No', 'Yes', 'No internet service']
TechSupport : ['No', 'Yes', 'No internet service']
StreamingTV : ['No', 'Yes', 'No internet service']
StreamingMovies : ['No', 'Yes', 'No internet service']
Contract : ['Month-to-month', 'One year', 'Two year']
PaperlessBilling : ['Yes', 'No']
PaymentMethod : ['Electronic check', 'Mailed check', 'Bank transfer (automatic)', 'Credit card (automatic)']
Churn : ['No', 'Yes']


## Target proportions

**Normalized** **value_counts** for **Churn** — share of Yes vs No (baseline class balance).

In [33]:
df['Churn'].value_counts(normalize=True)

Churn
No     0.73463
Yes    0.26537
Name: proportion, dtype: float64

# Data Exploration Summary

### Columns Description
- **customerID**: 7043 unique values  
  - ID of the customer
- **gender**: Male, Female  
  - Gender of the customer
- **SeniorCitizen**: 1, 0  
  - Senior citizen status (1 = Yes, 0 = No)
- **Partner**: Yes, No  
  - Has a partner or spouse
- **Dependents**: Yes, No  
  - Has dependents (children, elderly parents)
- **tenure**: Numeric  
  - Months the customer has stayed
- **PhoneService**: Yes, No  
  - Has phone service
- **MultipleLines**: Yes, No, No phone service  
  - Multiple phone lines
- **InternetService**: Fiber Optic, DSL, No  
  - Internet provider
- **OnlineSecurity**: Yes, No, No internet service  
  - Has online security
- **OnlineBackup**: Yes, No, No internet service  
  - Has online backup
- **DeviceProtection**: Yes, No, No internet service  
  - Has device protection
- **TechSupport**: Yes, No, No internet service  
  - Has tech support
- **StreamingTV**: Yes, No, No internet service  
  - Has streaming TV
- **StreamingMovies**: Yes, No, No internet service  
  - Has streaming movies
- **Contract**: Month-to-month, One year, Two year  
  - Type of contract
- **PaperlessBilling**: Yes, No  
  - Uses paperless billing
- **PaymentMethod**: Electronic check, Mailed check, Bank transfer (auto), Credit card (auto)  
  - Payment method
- **MonthlyCharges**: Numeric  
  - Amount charged monthly
- **TotalCharges**: Numeric  
  - Total amount charged
- **Churn**: Yes, No  
  - Whether the customer churned


---

### Data Insights

- **Shape**: (7043, 21)
- **Target Variable**: Churn (Binary classification problem)
- **Missing Values**: 11 missing values in `TotalCharges` correspond to customers with zero tenure (MonthlyCharges x 0 = 0)
- **Duplicates**: No duplicated records
- **Numerical Features**: tenure, MonthlyCharges, TotalCharges
- **Categorical Features**: Mix of binary and multi-class categorical variables
- **Binary Features**: Several Yes/No features that can be encoded during preprocessing
- **customerID**: Unique identifier, not useful for prediction (to be dropped)

---

### Important Observations

- **Class Imbalance**: ~27% churn vs ~73% non-churn